pkg/services/live/features/watch.go

```diff
@@ -0,0 +1,205 @@
+package features
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+
+	provisioning "github.com/grafana/grafana/pkg/apis/provisioning/v0alpha1"
+
+	"github.com/grafana/authlib/types"
+	"github.com/grafana/grafana-app-sdk/logging"
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/live"
+
+	"github.com/grafana/grafana/pkg/apimachinery/identity"
+	"github.com/grafana/grafana/pkg/services/apiserver"
+	"github.com/grafana/grafana/pkg/services/live/model"
+)
+
+// WatchRunner will start a watch task and broadcast results
+type WatchRunner struct {
+	publisher      model.ChannelPublisher
+	configProvider apiserver.RestConfigProvider
+
+	watchingMu sync.Mutex
+	watching   map[string]*watcher
+}
+
+func NewWatchRunner(publisher model.ChannelPublisher, configProvider apiserver.RestConfigProvider) *WatchRunner {
+	return &WatchRunner{
+		publisher:      publisher,
+		configProvider: configProvider,
+		watching:       make(map[string]*watcher),
+	}
+}
+
+func (b *WatchRunner) GetHandlerForPath(_ string) (model.ChannelHandler, error) {
+	return b, nil // all dashboards share the same handler
+}
+
+// Valid paths look like: {version}/{resource}[={name}]/{user.uid}
+// * v0alpha1/dashboards/u12345
+// * v0alpha1/dashboards=ABCD/u12345
+func (b *WatchRunner) OnSubscribe(ctx context.Context, u identity.Requester, e model.SubscribeEvent) (model.SubscribeReply, backend.SubscribeStreamStatus, error) {
+	// To make sure we do not share resources across users, in clude the UID in the path
+	userID := u.GetIdentifier()
+	if userID == "" {
+		return model.SubscribeReply{}, backend.SubscribeStreamStatusPermissionDenied, fmt.Errorf("missing user identity")
+	}
+	if !strings.HasSuffix(e.Path, userID) {
+		return model.SubscribeReply{}, backend.SubscribeStreamStatusPermissionDenied, fmt.Errorf("path must end with user uid (%s)", userID)
+	}
+
+	// While testing with provisioning repositories, we will limit this to admin only
+	if u.HasRole(identity.RoleAdmin) {
```

Comment:
sort term authz verification --- we should trust the watch command to do this, but lets start extra safe :)